Попов Захар

BlizPerfect opened the pull request:
@Folleach

Черновой вариант решения второй домашней работы.

Примеры работ с разными параметрами:
1. 
- Число прямоугольников = 100;
- Ширина прямоугольников от 30 до 70 пикселей;
- Высота прямоугольников от 20 до 50 пикселей.
![circularCloudLayouter](https://github.com/user-attachments/assets/7f83aa24-c3be-4c38-a042-6c67ea48f026)

2. 
- Число прямоугольников = 10000;
- Ширина прямоугольников от 30 до 70 пикселей;
- Высота прямоугольников от 20 до 50 пикселей.
![circularCloudLayouter](https://github.com/user-attachments/assets/fa3162c6-1428-465b-821e-fc748274c106)

3. 
- Число прямоугольников = 100;
- Ширина прямоугольников от 30 до 700 пикселей;
- Высота прямоугольников от 20 до 500 пикселей.
![circularCloudLayouter](https://github.com/user-attachments/assets/5cf95ab5-fb7e-457f-8842-530b99b4be0d)

4. 
- Число прямоугольников = 10;
- Ширина прямоугольников от 30 до 700 пикселей;
- Высота прямоугольников от 20 до 500 пикселей.
![circularCloudLayouter](https://github.com/user-attachments/assets/839af33d-aaba-4259-a111-2218710a45df)

5. 
- Число прямоугольников = 10;
- Ширина прямоугольников от 30 до 700 пикселей;
- Высота прямоугольников от 20 до 500 пикселей.
![circularCloudLayouter](https://github.com/user-attachments/assets/9f0ac478-2ca5-4522-b9b1-a2764f2f4493)

Принцип расположения прямоугольников:
Прямоугольники ставятся на окружности увеличивающегося радиуса с центром в точке, совпадающей с центром первого поставленного прямоугольника.

BlizPerfect committed:
Черновой вариант решения второго домашнего задания.

Folleach commented on cs/TagsCloudVisualization/TagsCloudVisualization.csproj:
```diff
@@ -0,0 +1,20 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
```
Нужно добавить этот `.csproj` в `tdd.sln`

Folleach replied:
Круто, что выделил!

Правда у меня упала сборка `TagsCloudVisualization.Tests`, вероятно, у тебя где-то был указан неявно атрибут?
```cs
[assembly: InternalsVisibleTo("TagsCloudVisualization.Tests")]
```

У меня ошибка:
```
Error CS0122 : 'CircularCloudLayouter' is inaccessible due to its protection level
```

А ещё в проекте `TagsCloudVisualization` теперь не нужны библиотеки для тестов, их нужно удалить

Folleach replied:
resolved

Folleach commented on cs/TagsCloudVisualization/CircularCloudLayouterTests.cs:
```diff
@@ -0,0 +1,85 @@
+
+        public void PutNextRectangle_ThrowsArgumentException_OnAnyNegativeOrZeroSize(int width, int height, int rectanglesCount)
+        {
+            SetUp();
```
Ой, вызывать SetUp при запуске теста NUnit умеет самостоятельно
https://docs.nunit.org/articles/nunit/writing-tests/attributes/setup.html

Folleach replied:
resolved

Folleach commented on cs/TagsCloudVisualization/CircularCloudLayouterTests.cs:
```diff
@@ -0,0 +1,85 @@
+
+            _circularCloudLayouterPainter.Save(_circularCloudLayouter.Rectangles.ToList());
+            return File.Exists(fileName);
+        }
```
Хорошим тоном будет удалять временные ресурсы, которые созданы во время работы тестов и больше ни для чего не нужны.  
В данном случае файл продолжит существовать после завершения тестов и уже никогда не будет удалён.

Предлагаю два варианта:
- Самостоятельно следить за жизненным циклом каждого файла
- Использовать временную директорию, в разных операционных системах она разная, в linux это `/tmp`

Folleach replied:
resolved

Folleach commented on cs/TagsCloudVisualization/CircularCloudLayouterWorker.cs:
```diff
@@ -0,0 +1,29 @@
+            var height = _random.Next(_minRectangleSize.Height, _maxRectangleSize.Height);
+            return new Size(width, height);
+        }
+    }
```
Правда ли для этой функциональности нужен экземпляр класса? Возможно получится сделать проще?

Folleach replied:
Со статическим классом действительно лучше!

С проверками правда, код, выглядит громоздко. У меня получилось сократить 58 строк в 27, сохранив читаемость

Folleach replied:
resolved

Folleach commented on cs/TagsCloudVisualization/CircularCloudLayouterTests.cs:
```diff
@@ -0,0 +1,85 @@
+    {
+        private CircularCloudLayouter _circularCloudLayouter;
+        private CircularCloudLayouterWorker _circularCloudLayouterWorker;
+        private CircularCloudLayouterPainter _circularCloudLayouterPainter;
```
Сейчас один класс с тестами содержит в себе всё - `Layouter` и `Painter`. Хотя одновременно они нужны не всегда.  
Лучше разделять тестовые классы по функциональности, которую они тестируют. Чтобы избежать большого количества полей.

Folleach replied:
resolved

Folleach commented on cs/TagsCloudVisualization/CircularCloudLayouterTests.cs:
```diff
@@ -0,0 +1,85 @@
+                var nextRectangleSize = _circularCloudLayouterWorker.GetNextRectangleSize();
+                Assert.That(_circularCloudLayouter.PutNextRectangle(nextRectangleSize), Is.TypeOf(typeof(Rectangle)));
+            }
+        }
```
Что проверяет этот тест?

Folleach replied:
resolved

Folleach commented on cs/TagsCloudVisualization/CircularCloudLayouterTests.cs:
```diff
@@ -0,0 +1,85 @@
+        public void PutNextRectangle_ThrowsArgumentException_OnAnyNegativeOrZeroSize(int width, int height, int rectanglesCount)
+        {
+            SetUp();
+            for (int i = 0; i < rectanglesCount; i++)
```
Если количество всегда равно единице, то стоит опустить цикл для упрощения теста.
Если один и тот же тест хочется запустить множество раз, вероятно, стоит взглянуть на атрибут [Repeat](https://docs.nunit.org/articles/nunit/writing-tests/attributes/repeat.html)

Folleach replied:
resolved

Folleach commented on cs/TagsCloudVisualization/CircularCloudLayouterWorker.cs:
```diff
@@ -0,0 +1,29 @@
+using System.Text;
+using System.Threading.Tasks;
+
+namespace TagsCloudVisualization
+{
```
Какой версией языка ты пользуешься? С C# 10 можно опустить блок с `namespace`, выглядит красивее :)

BlizPerfect replied:
Дело привычки!
Пока я не скован регламентом написания кода в какой-то конкретной команде, то я люблю, когда всегда явно указывается namespace. Это как с if - я ВСЕГДА ставлю скобки, даже если тело условия в одну строку - я считаю, что так в разы наглядней, хотя не могу не согласиться, что от этого код раздувается.

Folleach replied:
Хорошо! Мне не важно, какой стиль используется, главное чтобы он был везде одинаков :)

resolved

Folleach commented on cs/TagsCloudVisualization/CircularCloudLayouterWorker.cs:
```diff
@@ -0,0 +1,29 @@
+        {
+            _minRectangleSize = minRectangleSize;
+            _maxRectangleSize = maxRectangleSize;
+        }
```
А начиная с C# 12 можно использовать [primary constructor](https://learn.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-12#primary-constructors)

Folleach replied:
resolved

Folleach commented on cs/TagsCloudVisualization/CircularCloudLayouterPainter.cs:
```diff
@@ -0,0 +1,57 @@
+            _fileName = fileName;
+        }
+
+        public void Save(List<Rectangle> rectangles, int? paddingPerSide = null)
```
Как на счёт массивов?
Не хотелось бы их конвертировать в список, для того, чтобы вызвать метод `Save`

Folleach replied:
resolved

Folleach commented on cs/TagsCloudVisualization/CircularCloudLayouterPainter.cs:
```diff
@@ -0,0 +1,57 @@
+        public CircularCloudLayouterPainter(string fileName)
+        {
+            _fileName = fileName;
+        }
```
Почему `Painter` сохраняет именно в файл?  
Теоритически, нарисованное изображение могло быть отправлено по сети. Может быть можно декомпозировать эту логику?

Folleach replied:
resolved

Folleach commented on cs/TagsCloudVisualization/CircularCloudLayouterPainter.cs:
```diff
@@ -0,0 +1,57 @@
+                    bitmap.Save(_fileName, System.Drawing.Imaging.ImageFormat.Png);
+                }
+            }
+            Console.WriteLine($"Изображение сохранено как {_fileName}");
```
Этот вывод в консоль - сайд-эффект по отношению к `CircularCloudLayouterPainter`  
`Painter` мог бы находиться в библиотеке, которая подключается к другим проектам, у которых может быть свой структурированный вывод в консоль.
Но при сохранении изображения этот вывод будет нарушен.

Рекомендую придерживаться чистых функций, потому что с ними проще работать: они не меняют глобального состояния системы и выдают один и тот же результат для одинаковых входных параметров.

Folleach replied:
resolved

Folleach commented on cs/TagsCloudVisualization/CircularCloudLayouterPainter.cs:
```diff
@@ -0,0 +1,57 @@
+            if (rectangles.Count == 0)
+            {
+                return;
+            }
```
Правда ли пустое облако тегов не должно быть сохраено?
Из сигнатуры метода не понятно, что файла может не оказаться после вызова `Save`

Folleach commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,92 @@
+
+namespace TagsCloudVisualization
+{
+    internal class CircularCloudLayouter
```
В задаче предлагалось реализовать интерфейс, он действительно тут есть, но есть неявно.
А что, если мы захотим создать альтернативный `Layouter` для облака тегов? В этом случае, наличие интерфейса было бы очень кстати. Добавишь?

BlizPerfect replied:
Изначально я и хотел реализовывать интерфейс, но в интерфейсе нельзя указать требование, что бы был конструктор классов со специфичной сигнатурой, кроме как пустой конструктор. Потом подумал в сторону абстрактного класса, но в итоге решил через интерфейс реализовать.

Folleach replied:
Интерфейсы действительно не умеют описывать в себе конструктор. А это и не нужно, потому что набор методов, которые декларирует интерфейс может быть у чего угодно - структуры, класса, рекорда...

Более того, в зависимости от реализации интерфейса, эти конструкторы могут отличаться.
Например, конструктор для базы данных в памяти будет, скорее всего, пустым.
А конструктор для базы данных где-то на сервере будет иметь, как минимум, адрес этого сервера.

resolved

Folleach commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,92 @@
+
+namespace TagsCloudVisualization
+{
+    internal class CircularCloudLayouter
```
В классе `CircularCloudLayouter` сейчас сосредоточено много ответственности, можно ли что-нибудь вынести отсюда в отдельные классы?

Folleach replied:
С выделением дополнительных методов стало лучше.
Вероятно, можно было бы дальше улучшать читаемость, так как сейчас уровень вложенности достигает 3, но думаю, будет достаточно.

Из простого, разве что `if` можно было инвертировать :)
Из посложнее: сделать, чтобы `Circle` возвращал точки непрерывно, тогда `PutNextRectangle` мог выглядеть как-то так:

```cs
var result = circle
    .GetPoints()
    .Select(CreateRectangle)
    .Where(x => !IsIntersectionWithAlreadyPlaced(x))
    .First();
```

resolved

Folleach commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,92 @@
+            if (center.X < 0 || center.Y < 0)
+            {
+                throw new ArgumentException("Координаты центра не могут быть меньше нуля.");
+            }
```
Почему?
В декартовой системе координат вполне себе может существовать точка с отрицательными осями

А что будет, если кордината центра будет `(1, 1)`? Облако тегов будет в виде сектора окружности?

BlizPerfect replied:
Так как нигде ничего не говорилось про размер итоговой картинки, я решил поступить следующим образом - итоговое изображение всегда будет содержать в себе все прямоугольники. Даже если мы выставим центром точку (0,0) - на итоговом изображении этот центр сдвинется в центр непосредственно изображения.

Folleach replied:
Понял, звучит приемлемо.
Но не понятно, в чём тогда причина запрещать минусовые центры?

BlizPerfect replied:
Эта ошибочная логика возникла из-за того, что я сразу стал мыслить в координатах изображения, где левый верхний угол это (0,0) и отрицательных координат не может быть.

Folleach replied:
resolved

Folleach commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,92 @@
+        private Point[] GetCoordinatesOnEllipse(Point rectangleCenter, float radius)
+        {
+            var result = new Point[360];
+            var step = 1;
```
Правильно понимаю, что от 360 точек может понадобиться лишь 1? Возможно, в таком случае не нужно выделять массив на 360 элементов?

Folleach replied:
resolved

Folleach commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,92 @@
+
+                    int x = points[index].X - rectangleSize.Width / 2;
+                    int y = points[index].Y - rectangleSize.Height / 2;
+                    var location = new Point(x, y);
```
Рекомендую разнообразные математические (и не только) операции выносить в методы расширения или другие статические классы. Тогда их использование должно стать проще, как и чтение кода.

А ещё, их можно будет тестировать!

Folleach replied:
resolved

Folleach commented on cs/TagsCloudVisualization/CircularCloudLayouterTests.cs:
```diff
@@ -0,0 +1,85 @@
+namespace TagsCloudVisualization
+{
+    [TestFixture]
+    internal class CircularCloudLayouterTests
```
Тесты нужны лишь на этапе разработки. При публикации кода куда-либо, тесты не нужны. Но сейчас они будут скомпилированы в `dll` с основным кодом.
Будет здорово, если бы тесты можно было отделить от основного кода. Исправишь?

Folleach replied:
resolved

Folleach commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,92 @@
+        private float _radius = 2.0f;
+        private Random _random = new Random();
+        private readonly List<Rectangle> _rectangles = new List<Rectangle>();
+        public IReadOnlyList<Rectangle> Rectangles => _rectangles.AsReadOnly();
```
Давай не будем делать прямоугольники публичными?  
Метод `PutNextRectangle` уже возвращает прямоугольники, этого должно хватить

Folleach replied:
resolved

Folleach commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,92 @@
+            var result = new Point[360];
+            var step = 1;
+
+            for (int angle = 0; angle < 360; angle += step)
```
Почему в некоторых местах используется `var`, а в некоторых явное задание типа?

BlizPerfect replied:
Это мне когда студия сама подсказывает код, я там забываю поменять иногда тип данных, но всегда сам использую var)

Folleach replied:
Понял. Но как и писал выше - хотелось бы видеть единобразный стиль во всём коде

BlizPerfect replied:
Я полностью с Вами согласен - тоже считаю, что всё должно быть единообразно.

Конкретно с var в циклах у меня забавная ситуация:
При решении задач на Leetcode я сначала руками всё писал через var, потом начал принимать подсказки студии, но менять int в цикле на var, но в итоге сдался и перестал это замечать)

Folleach replied:
Возможно Visual Studio можно настраить, чтобы она давала правильные подскази? :)

Не знаю как в ней, но в rider для этого есть настраиваемые live templates
![image](https://github.com/user-attachments/assets/2f862c60-7178-41d4-ad0b-e413519e3ba1)

В vs code это называется [snippets](https://code.visualstudio.com/docs/editor/userdefinedsnippets)
В visual studio вроде тоже [snippets](https://learn.microsoft.com/en-us/visualstudio/ide/code-snippets?view=vs-2022)

Folleach reviewed (changes requested):
### Резюмируя

В целом для предпроверки - нормально, но код явно ещё нужнается в причёсывании. Хотелось бы видеть больше декомпозии на отдельные компонены, чтобы гига-методы с уровнем вложенности 4 стали проще!

Кроме того, хочется больше тестов, но думаю, ко второму дедлайну они будут доделаны. Простор для тестов тут, на самом деле, огромный. Не стесняйся тестировать даже то, что выходит за рамки основных классов: экстеншены, хелперы и всякие другие мелочи, которые есть.
Но как минимум, ожидаю увидеть тесты на все требования к раскладке

И не забывай обращать внимание на разнообразные предупреждения IDE, они помогают улучшить общую красоту кода ^_^

Про git могу сказать, круто, что теперь ветка отдельная, но название не несёт в себе какой-либо полезной информации. Хотелось бы, чтобы оно было понятнее.
На практике в названии ветки часто содержиться номер задачи или никнейм автора, а потом пара ключевых слов, о чём ветка, чтобы можно было их различать:
- task-1234/circular-visualizer
- BlizPerfect/render-engine

BlizPerfect committed:
TagsCloudVisualization.csproj явно добавлен в tdd.sln.

BlizPerfect committed:
Класс CircularCloudLayouterWorker теперь является статичным.

Класс CircularCloudLayouterWorker был переписан как статичный для избежания создания экземпляра класса.

BlizPerfect committed:
Метод Save у CircularCloudLayouterPainter теперь принимает IList<Rectangle>

BlizPerfect committed:
Добавлен интерфейс ICircularCloudLayouter.

1. Добавлен интерфейс ICircularCloudLayouter.
2. Класс CircularCloudLayouter теперь реализует интерфейс ICircularCloudLayouter.

BlizPerfect committed:
Рефакторинг класса CircularCloudLayouter.

1. Рефакторинг класса CircularCloudLayouter.
2. Выделение сущности Circle, отвечающей за полчение координат на окружности.

BlizPerfect committed:
Разделён функционал отрисовки прямоугольников и сохранения изображения

BlizPerfect committed:
Добавлена проверка входных данных у CircularCloudLayouterWorker.

BlizPerfect committed:
Общий рефакторинг кода.

1. Добавление новых тестов.
2. Общий рефакторинг кода.

BlizPerfect committed:
Update README.md

BlizPerfect committed:
Update README.md

Folleach commented on cs/TagsCloudVisualization.Tests/CircularCloudLayouterMainRequirementsTests.cs:
```diff
@@ -0,0 +1,138 @@
+            }
+        }
+
+        [TestCase(1.1, ExpectedResult = true)]
+        [Repeat(10)]
+        public bool ShouldPlaceRectanglesInCircle(double goodAspectRatio)
+        {
+            var boundingBoxSize = GetBoundingBoxSize();
+            var actualAspectRatio =
+                (double)Math.Max(boundingBoxSize.Width, boundingBoxSize.Height)
+                / Math.Min(boundingBoxSize.Width, boundingBoxSize.Height);
+            return actualAspectRatio <= goodAspectRatio;
+        }
```
А если прямоугольники образуют квадрат?

Folleach replied:
Хорошо, квадрат теперь не проходит, но разница очень небольшая. Возможно это из-за специфической логики с `grid`, в которую я особо не вдавался из-за её сложности

resolved

Folleach commented on cs/TagsCloudVisualization.Tests/CircularCloudLayouterMainRequirementsTests.cs:
```diff
@@ -0,0 +1,138 @@
+            var actualAspectRatio =
+                (double)Math.Max(boundingBoxSize.Width, boundingBoxSize.Height)
+                / Math.Min(boundingBoxSize.Width, boundingBoxSize.Height);
+            return actualAspectRatio <= goodAspectRatio;
```
Числа лучше сравнивать специальными методами тестового фреймворка. Тогда при падении теста будет видно значение, а не абстрактное
```
Expected: True
Actual: False
```
В первом домашнем задании с этим боролись как раз

```cs
actualAspectRatio.Should().BeLessThan(goodAspectRatio);
```

Folleach replied:
resolved

Folleach commented on cs/TagsCloudVisualization.Tests/CircularCloudLayouterMainRequirementsTests.cs:
```diff
@@ -0,0 +1,138 @@
+            return actualAspectRatio <= goodAspectRatio;
+        }
+
+        [TestCase(0.1, ExpectedResult = true)]
+        [Repeat(10)]
+        public bool ShouldPlaceRectanglesNearCenter(double percentFromBoundingBoxMaxSize)
+        {
+            var centerX = _rectangles.Average(r => r.Left + r.Width / 2.0);
+            var centerY = _rectangles.Average(r => r.Top + r.Height / 2.0);
+            var actualCenter = new Point((int)centerX, (int)centerY);
+
+            var distance = Math.Sqrt(Math.Pow(actualCenter.X - _center.X, 2)
+                                     + Math.Pow(actualCenter.Y - _center.Y, 2));
+
+            var boundingBoxSize = GetBoundingBoxSize();
+
+            var maxAllowedDistance = Math.Max(boundingBoxSize.Width, boundingBoxSize.Height)
+                                     * percentFromBoundingBoxMaxSize;
+
+            return maxAllowedDistance >= distance;
+        }
```
Центр в точке `(400, 400)`
Тестируемые прямоугольники:
```cs
_rectangles =
[
    new Rectangle(new Point(1000400, 400), new Size(1, 1)),
    new Rectangle(new Point(-999600, 400), new Size(1, 1))
];
```

Тест считает, что прямоугольники расположены прямо около центра: `distance = 0`. Но это не так.
Точно ли тогда тест проверяет, что прямоугольники должны быть близки к центру?

Folleach replied:
\+ ожидаемое значение динамически лучше не высчитывать. Если у тебя есть заданные параметры для генерации, стоит отталкиваться от них и выбрать допустимый предел

Folleach replied:
Тест всё ещё проходит с примером выше.  
Но, вероятно, одновременное прохождение тестов `ShouldPlaceRectanglesInCircle` и `ShouldPlaceCenterOfMassOfRectanglesNearCenter` может означать, что прямоугольники распологаются около центра.

resolved

Folleach commented on cs/TagsCloudVisualization.Tests/CircularCloudLayouterMainRequirementsTests.cs:
```diff
@@ -0,0 +1,138 @@
+        }
+
+
+        [TestCase(0.55, ExpectedResult = true)]
+        [Repeat(10)]
+        public bool ShouldPlaceRectanglesTightly(double tight)
+        {
+            var boundingBoxSize = GetBoundingBoxSize();
+            var boundingBoxArea = boundingBoxSize.Width * boundingBoxSize.Height;
+            var filledArea = _rectangles.Sum(r => r.Width * r.Height);
+            return ((double)filledArea / boundingBoxArea) >= tight;
+        }
```
Ой, тут, кажется, тоже квадрат пройдёт тест

Folleach replied:
Видимо этот тест был убран в угоду `ShouldPlaceCenterOfMassOfRectanglesNearCenter`

resolved

Folleach commented on cs/TagsCloudVisualization.Tests/CircularCloudLayouterPainterTests.cs:
```diff
@@ -0,0 +1,38 @@
+                () => CircularCloudLayouterPainter.Draw(new List<Rectangle>()));
+        }
+
+        [TestCase(null, ExpectedResult = true)]
+        [TestCase(100, ExpectedResult = true)]
+        public bool Draw_CalculatesImageSizeCorrectly(int? padding)
+        {
+            var correctPadding = padding ?? _defaultPadding;
```
Дефолтное значение удобно применять в публичных интерфейсах (например библиотек), чтобы лишний раз не заморачиваться над передачей приемлемых по умолчанию параметров.  
Как сейчас сделано в
```cs
Bitmap Draw(IList<Rectangle> rectangles, int? paddingPerSide = null)
```
Но в тестах это лишнее. Можно `null` в `TestCase` заменить на 10 - будет нагляднее

Folleach commented on cs/TagsCloudVisualization.Tests/RectangleExtensions.cs:
```diff
@@ -0,0 +1,18 @@
+{
+    internal static class RectangleExtensions
+    {
+        public static string ToFormatedString(this Rectangle rectangle)
+        {
+            return $"X={rectangle.X}, Y={rectangle.Y}, Width={rectangle.Width}, Height={rectangle.Height}";
+        }
```
Вроде `.ToString()` умеет выводить содержимое прямоугольника  
https://github.com/dotnet/runtime/blob/main/src/libraries/System.Drawing.Primitives/src/System/Drawing/Rectangle.cs#L340-L343

Folleach replied:
resolved

Folleach commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,69 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Drawing;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
```
Заметил много ненужных `using` по всем классам, давай удалять лишнее?

Folleach replied:
resolved

Folleach commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,69 @@
+{
+    internal class CircularCloudLayouter(Point center) : ICircularCloudLayouter
+    {
+        private Circle _arrangementСircle = new Circle(center);
+        private Random _random = new Random();
```
Если ссылка никогда не модифицируется - можно смело делать её `readonly`

Folleach replied:
В других местах тоже стоит исправлять комментарии

BlizPerfect committed:
Из проекта TagsCloudVisualization удалены библиотеки для тестирования.

BlizPerfect committed:
В проекте TagsCloudVisualization явно указан атрибут видимости internal классов для тестов.

BlizPerfect committed:
Рефакторинг тестов на проверку основным требованиям.

1. Тесты ShouldPlaceRectanglesInCircle и ShouldPlaceRectanglesTightly были заменены на более точный тест ShouldPlaceRectanglesInCircle, обьединяющий их функциональность.
2. ShouldPlaceRectanglesNearCenter переименован в ShouldPlaceCenterOfMassOfRectanglesNearCenter, так как предыдущее название вводило в заблуждение по функциональности теста.
3. Рефакторинг тестов.

BlizPerfect committed:
Удаление неиспользуемых using.

BlizPerfect committed:
CircularCloudLayouterWorker приведён в более приятный глазу вид.

BlizPerfect committed:
Код приведён к единообразию использованя var.

BlizPerfect committed:
Merge branch 'Dev' of https://github.com/BlizPerfect/tdd into Dev

Folleach commented on cs/TagsCloudVisualization.Tests/CircularCloudLayouterMainRequirementsTests.cs:
```diff
@@ -0,0 +1,168 @@
+
+        private bool[,] GetOccupancyGrid(int gridSize, double maxRadius, double step)
+        {
+            var result = new bool[gridSize, gridSize];
```
Массив байт 1000x1000 это гарантированное попадание в [Large Object Heap](https://learn.microsoft.com/en-us/dotnet/standard/garbage-collection/large-object-heap)
> If an object is greater than or equal to 85,000 bytes in size, it's considered a large object

Большие объекты дольше создаются, дольше удаляются (причём только при сборке 2-го поколения) и занимают много места в оперативной памяти. Подробнее можно почитать по ссылке выше.

Поэтому рекомендуется по возможности их избегать.

В данном случае, что было бы, если облако тегов содержало 2 диаметрально противоположных прямоугольнка на дальних координатах?  
Им может банально не хватить оперативной памяти компьютера, чтобы "закрасить" этот `grid`.

Альтернативный вариант, который тут можно было реализовать: подсчёт площадей.
Мы можем легко посчитать сумму всех площадей прямоугольников: O(n) по времени, O(1) по памяти.
А также можем найти радиус описанной окружности всех прямоугольников. Что позволит посчитать площадь.
Отношение этих двух чисел будет показывать коэффициент плотности располагаемых прямоугольников.

Folleach commented on cs/TagsCloudVisualization.Tests/RectangleExtensions.cs:
```diff
@@ -0,0 +1,16 @@
+                Math.Abs(rectangle.X - point.X), Math.Abs(rectangle.X + rectangle.Width - point.X));
+            var dy = Math.Max(
+                Math.Abs(rectangle.Y - point.Y), Math.Abs(rectangle.Y + rectangle.Height - point.Y));
+            return Math.Sqrt(dx * dx + dy * dy);
```
Функция встречается трижды в коде - хороший кандидат на выделение в метод

Folleach replied:
Функция встречается трижды в коде - хороший кандидат на выделение в метод
